File: backend/API/PY_bots/Business/tg_business_bot/Models/Restaurant.py
from pydantic import BaseModel
from typing import List, Optional

from Models.NotificationGetter import NotificationGetter
from Models.Menu import Menu
from Models.Table import Table
from Models.GeoCoordinates import GeoCoordinates

class Restaurant(BaseModel):
    id: Optional[int]
    owner_id: int
    notification_getter: Optional[NotificationGetter]
    name: str
    coordinates: GeoCoordinates
    menu: Optional[Menu]
    tables: Optional[List[Table]]

    def __init__(self,
                 owner_id: int,
                 name: str,
                 coordinates: GeoCoordinates,
                 id: Optional[int] = None,
                 notification_getter: Optional[NotificationGetter] = None,
                 menu: Optional[Menu] = None,
                 tables: Optional[List[Table]] = None):
        super().__init__(
            id=id,
            owner_id=owner_id,
            notification_getter=notification_getter,
            name=name,
            coordinates=coordinates,
            menu=menu,
            tables=tables
        )

File: Domain/Models/Restaurant.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using Domain.DataBaseAccess;
using Domain.Logic;
using Microsoft.EntityFrameworkCore;

namespace Domain.Models;

public class Restaurant : ITableDataBase, ITakeRelatedData, IDeleteRelatedData
{
    public long? Id { get; set; }
    [JsonPropertyName("owner_id")] public long OwnerId { get; set; }
    public string Name { get; set; }
    [NotMapped] public GeoCoordinates? Coordinates { get; set; }
    [JsonPropertyName("notification_getter")] [NotMapped] public NotificationGetter? NotificationGetter { get; set; }
    [NotMapped] public Menu? Menu { get; set; }
    [NotMapped] public List<Table>? Tables { get; set; }

    public Restaurant(){}
    public Restaurant(long? id, long ownerId, string name, GeoCoordinates? coordinates, NotificationGetter? notificationGetter,
        Menu? menu, List<Table>? tables)
    {
        Id = id;
        OwnerId = ownerId;
        Name = name;
        Coordinates = coordinates;
        NotificationGetter = notificationGetter;
        Menu = menu;
        Tables = tables;
    }

    public async Task TakeRelatedData(ApplicationDbContext context)
    {
        Menu = await context.Menus
            .FirstOrDefaultAsync(m => m.RestaurantId == Id);
        if (Menu != null)
            await Menu.TakeRelatedData(context);

        NotificationGetter = await context.NotificationGetters.FirstOrDefaultAsync(n => n.RestaurantId == Id);

        Tables = await context.Tables
            .Where(t => t.RestaurantId == Id)
            .ToListAsync();

        Coordinates = await context.GeoCoordinates.FirstOrDefaultAsync(c => c.RestaurantId == Id);
    }

    public void DeleteRelatedData(ApplicationDbContext context)
    {
        var menu = context.Menus.FirstOrDefault(m => m.RestaurantId == Id);
        if (menu != null)
        {
            context.Menus.Remove(menu);
            context.SaveChanges();
            menu.DeleteRelatedData(context);
        }

        foreach (var table in context.Tables.Where(t => t.RestaurantId == Id))
        {
            context.Tables.Remove(table);
            context.SaveChanges();
        }

        var coordinates = context.GeoCoordinates.FirstOrDefault(n => n.RestaurantId == Id);
        if (coordinates != null)
        {
            context.GeoCoordinates.Remove(coordinates);
            context.SaveChanges();
        }
    }
}